polus-precompute-volume-plugin/src/utils.py

```diff
@@ -0,0 +1,338 @@
+                                            y=(ysplits[y], ysplits[y+1]),
+                                            x=(xsplits[x], xsplits[x+1]),  
+                                            z=(zsplits[z], zsplits[z+1]))
+                            logger.info("Loaded subvolume (YXZ) {}-{}__{}-{}__{}-{}".format(start_y, end_y,
+                                                                                            start_x, end_x,
+                                                                                            start_z, end_z))
+
+                            ids = np.unique(volume)
+                            if (ids == [0]).all():
+                                continue
+                            else:
+                                ids = np.delete(ids, np.where(ids==0))
+                                with ThreadPoolExecutor(max_workers=8) as executor:
+                                    executor.submit(create_plyfiles(subvolume = volume,
+                                                                    ids=ids,
+                                                                    temp_dir=temp_dir,
+                                                                    start_y=start_y,
+                                                                    start_x=start_x,
+                                                                    start_z=start_z,
+                                                                    totalbytes=totalbytes))
+                                    all_identities = np.append(all_identities, ids)
+
+                executor.shutdown(wait=True)
+
+                logger.info("\n Generating Volumes ...")
+                logger.info("Saved Encoded Volumes for Scale {}".format(num_scales))
+
+                # Build rest of the pyramid from encoded volumes in highest resolution
+                for higher_scale in reversed(range(0, num_scales)):
+
+                    inputshape = np.ceil(bfshape/(2**(num_scales-higher_scale-1))).astype('int')
+                    scale_directory = os.path.join(output_image, str(higher_scale+1))
+
+                    logger.info("Saved Encoded Volumes for Scale {}".format(higher_scale))
+
+                    if not os.path.exists(highest_res_directory):
+                        os.makedirs(scale_directory, exist_ok=True)
+                    ngvol.get_rest_of_the_pyramid(directory=scale_directory, input_shape = inputshape, chunk_size=chunk_size,
+                                                 datatype=datatype)
+
+                logger.info("\n Generate Progressive Meshes for segments ...")
+                # concatenate and decompose the meshes in the temporary file for all segments
+                all_identities = np.unique(all_identities).astype('int')
+                with ThreadPoolExecutor(max_workers=4) as executor:
+                    futuresvariable = [executor.submit(concatenate_and_generate_meshes, 
```

Comment:
I thought one of the goals of the recent effort was to not loop over the data multiple times. It appears as though the current structure of the code still does that. It is taking a different form than the recursive strategy, but you still have two different sections of the code that loop over the data separately. Am I interpreting this incorrectly?

Reply:
So it reads encoded data, but with each loop the data is blurred and the number of files it reads reduces anywhere from 2-8 times with every loop.  It also reads a byte stream instead of an image with every iteration. 
What I can do is add in extra if statements so that if there is only one file in the directory and it still hasn't reached level zero, then it can all be loaded into memory and blur the images for the rest of the pyramid.

If its just better to go through the input twice, I can change back to include the recursive function. I thought this method would be better for extremely large inputs.

Reply:
That's true for the image, but you're calculating the mesh outside of the for loop, correct? That means your loading the data from the binary files two times: once for generating the segmentation and another time for creating the mesh, right?

If you're going to read data directly from the binary files, why not just use the recursive approach to generating the segmentation pyramids, then load those binary files to calculate the meshes?